### PR TITLE
Use implicit/explicit extension as syntax

### PR DESCRIPTION
**Pull request by jcouv**
Yesterday's roles WG took a hard look at the proposed syntax and updated the proposal to use `explicit extension` (for "role") and `implicit extension` (for "extension"). An `implicit` or `explicit` keyword is required.
Also, the underlying type was also taken out of the base list and follows the keyword `for`: `implicit extension X for U : BaseExtension, Interface`.

In terms of implementation (syntax model), note that the `implicit` and `explicit` keywords are not stored in the modifiers list. This is following the existing model of operators. So we have pairs of keywords, like `implicit extension`, `explicit extension` like we do `implicit operator` (or `record struct` FWIW).

Relates to test plan https://github.com/dotnet/roslyn/issues/66722
Speclet https://github.com/jcouv/csharplang/blob/roles-spec/proposals/roles.md

### CODE REVIEW

**Comment by jcouv** [comment]
@AlekseyTs @333fred @dotnet/roslyn-compiler for review. Thanks

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:1447]: Given that we've changed this for most other type declarations in the primary-ctor branch, we probably want to allow the semi-colon body here too. Worth leaving a prototype to followup after primary-ctors are merged.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3160]: I'd like to see some tests covering:
* Extensions named `extension`
* Explicitly-implemented implicit and explicit operators implemented:
    * From an interface named `extension`.
    * From an interface in a namespace named `extension`.
* Underlying types of contextual keywords, such as `dynamic` and `var.
* `for` after the base list.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 4).

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3506]: It's optional for partial scenarios based on last week's discussion

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1494]: Yes, I'm waiting for the code to flow first

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1487]: I didn't follow. What syntax would you like to improve recovery for?

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:1447]: There's already a PROTOTYPE comment below

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1644]: I didn't want to do that at this point as it's not essential to unblock binding work. But I can add a PROTOTYPE comment.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:13]: Yup. Will do that after the PR is reviewed.

File: src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Generated/CSharp.Generated.g4:324]: ick.  it would be nice if we could update our grammar generator to understand contextial identifiers (like 'extension').  actually.  does it not understand that?  it seems to work for other constructs (like 'from_clause').  so i would expect to see `'extension'` here not `syntax_token`.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3500]: this seems wrong.  the actuak kind should be ExtensionKeyword, not hte contextual kind. contextual kind is for when we have an identifier and don't know waht it is.  but for things like `implicit/explicit extension` we know it's a keyword and shioudl convert to that.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1412]: did higher up logic already verify this is an extension and not an operator?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1434]: this name seems useless (esp. since it also handles records :)).  How about just ParseTypeDeclaration?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1447]: maybe `typeKeyword` with a comment on it like `// class/struct/interface/record/extension`

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1446]: comment what these might be?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1447]: acutally, definitely comment.  looking below it's not clear what all three are for.

Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1447]: do we need all 3 at the same time?  don't you only have at most one of these?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1475]: consider calling this extensionUnderlyingType to make it clear this is just for extensions.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1630]: oh.  we do convert this to a keyword.  ok, then the Syntax.xml is wrong.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1629]: no need to convert this.  implicit/explicit *are* keywords.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1651]: it's confusing to me that we have mainKeyword and firstKeyword.  especially since we clearly see that mainKeyword here is the ClassKeyword.  but we then do *not* pass that in for `ClassDeclaration( ... keyword: ... )`

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2126]: how is there a compat break here?  this feels like `yield return`.  Because of implicit/explicit, we always know this is an extension and can safely parse it out.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3506]: not sure how i feel about a whole node for this.  but i can live with it :)

**Review by CyrusNajmabadi — COMMENTED** [review]
done with pass.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1471]: The `for Type`. IE, if I said `class X for Type`, I'd like to see us throw the `for Type` into skipped trivia and say "expected `:`". Or perhaps just eat the `for` as if it was a `:` instead.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs]: TODO2?

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3191]: Let's have a version with `implicit`/`explicit` conversion operators, as that should appear nearly identical to an extension type until we see the `.`.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1471]: Thanks for clarifying.
Perhaps. Will add a PROTOTYPE comment

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1447]: We need to discriminate based on the kind of type (class/struct/interface/record/extension). Keeping track of which of the two keywords is the "main" keyword achieves that (without having to introduce many boolean flags or an enum).

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1434]: records are classes or structs, so arguably they fit in the list.
Renaming to ParseTypeDeclaration results in an error (identical overloads).
ParseTypeDeclarationButEnumOrDelegate? ;-)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:2126]: Yup, implicit/explicit are now required.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1412]: Yes, calls to `ParseTypeDeclaration` are both preceded by attempts to parse operators

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3191]: I'm not sure what you mean. For conversion operators, the `operator` keyword must come right after `implicit`/`explicit`. The first opportunity for an `extension` token is after `operator`, so it wouldn't look much like an extension.
Added tests to illustrate.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1479]: We can still reorder these two parsing sections now though 🙂.

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml
Comment (333fred) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3506]: I would agree, this doesn't seem like it needsa node. Feels like it should be a `For` token and an underlying type node.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3414]: Sorry, this is still not what I'm asking for. I'm asking for an implicit and explicit conversion, explicitly implemented, from a type named extension and from a type in a namespace named extension.

```cs
    static explicit extension<extension>.operator extension(int i)
    static explicit extension.C<C>.operator C(int i)
```

**Review by 333fred — COMMENTED** [review]
Done review pass

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (CyrusNajmabadi) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3414]: you are bad, and you should feel bad :D.

Ugh....  So, i'm ok with a convo about a breaking this case (outside of this PR).  Seems awful to try to continue supporting that :D

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3506]: We have a node for name-equals. I think the same model applies here, since the pair (for+type) is optional, but the individual parts are not.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3414]: I'm going to leave this as a prototype comment. Not worth delaying the core of the feature. We can discuss later.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1460]: >mainKeyword = firstKeyword; [](http://example.com/codeflow?start=16&length=27)

Consider explicitly assigning `null` to ```secondKeyword``` for clarity.

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3414]: FWIW, I could not find a scenario where the above code is legal from a binding standpoint. Can you see one?

**Comment by jcouv** [comment]
@333fred @AlekseyTs for review. Thanks

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1673]: >firstKeyword is not null [](http://example.com/codeflow?start=43&length=24)

Assert that it is the ``mainKeyword``` instead?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1694]: >firstKeyword is not null [](http://example.com/codeflow?start=43&length=24)

Assert that it is the ``mainKeyword``` instead?

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:1718]: >firstKeyword is not null [](http://example.com/codeflow?start=43&length=24)

Assert that it is the ``mainKeyword``` instead?

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3499]: >Keyword [](http://example.com/codeflow?start=17&length=7)

Rename to "ExtensionKeyword"?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3499]: I think `Keyword` is preferrable to match `TypeDeclarationSyntax`.

**Comment by AlekseyTs** [comment]
>         UsingTree(text, options: TestOptions.Regular10,

11?

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:70 in 4c3c4a3. [](commit_id = 4c3c4a3227fbc34b7027da2f7f531ccd27666f5f, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         UsingTree(text, options: TestOptions.Regular10,

In other similar places too

---
In reply to: [1450595127](https://github.com/dotnet/roslyn/pull/66919#issuecomment-1450595127) [](http://example.com/codeflow?ancestors=1450595127)

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:70 in 4c3c4a3. [](commit_id = 4c3c4a3227fbc34b7027da2f7f531ccd27666f5f, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         UsingTreeWithCSharpNext(text);

I assume we expect an error during binding. Consider adding a PROTOTYPE comment 

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:2668 in 4c3c4a3. [](commit_id = 4c3c4a3227fbc34b7027da2f7f531ccd27666f5f, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     static extension extension.operator+(extension one, extension other) => throw null;

Did you mean to use `implicit`/`explicit` here instead?

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3196 in 4c3c4a3. [](commit_id = 4c3c4a3227fbc34b7027da2f7f531ccd27666f5f, deletion_comment = False)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 10)

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3506]: From offline discussion, Fred was okay with the current structure to handle optionality, but I captured a bullet in test plan to highlight/discuss this choice in API review. Thanks

File: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:3414]: https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKAJgjAFMBnABwgGNSACUgDwBdSZyoB7GAbgILoAzPVhsATgDNaDAJIAeACoA+egHcAFqXENF9EPQUqCAbwL0L9YfQjByLcbRZXsANkZNKAGyg0ozrkptCBYucVEYFgAKPSgASn58AF8BfCx6AGF9Q3kM5VNzS3Q3D29ffw82Dm4YADoFPNrA4NDwsSis+MKLM3xLfqsAdnpsRP6U5KA===.

**Comment by jcouv** [comment]
>         UsingTreeWithCSharpNext(text);

Yes, this will be a binding error. I have a test for this in follow-up PR.

---
In reply to: [1450623362](https://github.com/dotnet/roslyn/pull/66919#issuecomment-1450623362) [](http://example.com/codeflow?ancestors=1450623362,1450623362)

---
Refers to: src/Compilers/CSharp/Test/Syntax/Parsing/RoleParsingTests.cs:2668 in 3ce5939. [](commit_id = 3ce593957788c3267464bf1cb5d8bc40470fc0b8, deletion_comment = False)

**Pull request merged**
